### PR TITLE
[Validator] check for file existence lazily in loaders

### LINKED ISSUE #39945: [Validator] Unnecasarry validation metadata disk reads
**Symfony version(s) affected**: >= 4.4

**Description**  

Looks like [this code](https://github.com/symfony/validator/blob/5.x/Mapping/Loader/FileLoader.php#L37-L47) causes disk reads, when rendering a form, this code is called, even when validation does not take place:

```
#0  Symfony\Component\Validator\Mapping\Loader\FileLoader->__construct() called at [vendor/symfony/symfony/src/Symfony/Component/Validator/ValidatorBuilder.php:326]
#1  Symfony\Component\Validator\ValidatorBuilder->getLoaders() called at [vendor/symfony/symfony/src/Symfony/Component/Validator/ValidatorBuilder.php:352]
#2  Symfony\Component\Validator\ValidatorBuilder->getValidator() called at [app/cache/dev/ContainerVvqCaGq/appAppKernelDevDebugContainer.php:6813]
#3  ContainerVvqCaGq\appAppKernelDevDebugContainer->getValidatorService() called at [app/cache/dev/ContainerVvqCaGq/getForm_TypeExtension_Form_ValidatorService.php:9]
#4  require(app/cache/dev/ContainerVvqCaGq/getForm_TypeExtension_Form_ValidatorService.php) called at [app/cache/dev/ContainerVvqCaGq/appAppKernelDevDebugContainer.php:2818]
#5  ContainerVvqCaGq\appAppKernelDevDebugContainer->load() called at [app/cache/dev/ContainerVvqCaGq/getForm_RegistryService.php:228]
#6  ContainerVvqCaGq\appAppKernelDevDebugContainer->{closure}() called at [vendor/symfony/symfony/src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php:56]
#7  Symfony\Component\Form\Extension\DependencyInjection\DependencyInjectionExtension->getTypeExtensions() called at [vendor/symfony/symfony/src/Symfony/Component/Form/FormRegistry.php:120]
#8  Symfony\Component\Form\FormRegistry->resolveType() called at [vendor/symfony/symfony/src/Symfony/Component/Form/FormRegistry.php:94]
#9  Symfony\Component\Form\FormRegistry->getType() called at [vendor/symfony/symfony/src/Symfony/Component/Form/FormRegistry.php:127]
#10 Symfony\Component\Form\FormRegistry->resolveType() called at [vendor/symfony/symfony/src/Symfony/Component/Form/FormRegistry.php:94]
#11 Symfony\Component\Form\FormRegistry->getType() called at [vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php:58]
#12 Symfony\Component\Form\FormFactory->createBuilder() called at [vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php:30]
#13 Symfony\Component\Form\FormFactory->create() called at [vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php:324]
```

Looks like `ValidatorBuilder::getLoaders()` only creates loaders and constructors fire disk reads, even if `FileLoader::loadClassMetadata()` is not called.

**Possible Solution**  
Check `is_file()`, `is_readable()`, `stream_is_local()` only when validation is executed (`FileLoader::oadClassMetadata()` is called). 
This can be further reduced with caching, so it des not check files on each HTTP request.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #39945
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php
Comment (xabbuh) [inline src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: I would prefer not to call the constructor here but move the logic that is located there in the base class into a dedicated method that we can call here then.

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: that'd mean opening an API for that :(

Comment (xabbuh) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: We can mark the method as internal.

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: we can, but what's the issue with calling the constructor?

Comment (stof) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: IIRC, calling our constructors as non-constructor methods is not covered by our BC policy.

Here, you are still opening an API without telling it.

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: I'm not opening anything, the constructor of the base class won't change in a non-BC way

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: In case that's what is bothering you, the parent constructor is still going to be called only once.

Comment (xabbuh) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: That’s okay, it just doesn’t feel right to me to call the constructor at a later time.

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Mapping/Loader/XmlFileLoader.php:187]: Not sure why though. ProxyManager does it all the time for example. Lazy initialization is legit concern.

**Review by fancyweb — APPROVED** [review]
What if someone extends XmlFileLoader and expects the $file property to be "validated" after the instantiation (which is not the case anymore)?

**Comment by nicolas-grekas** [comment]
> What if someone extends XmlFileLoader and expects the $file property to be "validated" after the instantiation (which is not the case anymore)?

That's not the main domain logic of this class, we don't have to preserve such behaviors. And it's not like we're removing the check since we're just moving it to were it's actually needed.

**Review by ogizanagi — APPROVED** [review]

**Comment by fancyweb** [comment]
Let's wait for @umpirsky feedback on the original issue before merging I guess :-)

**Review by chalasr — APPROVED** [review]
Let's move forward, we can revisit later if needed.

**Pull request merged**
